provider/pkg/resources/cfn_custom_resource.go

```diff
@@ -0,0 +1,681 @@
+					TypeSpec:    pschema.TypeSpec{Type: "boolean"},
+				},
+			},
+			Required: []string{"physicalResourceId", "data", "stackId", "serviceToken", "bucket", "resourceType", "noEcho"},
+		},
+		InputProperties: map[string]pschema.PropertySpec{
+			"bucketName": {
+				Description: "The name of the S3 bucket to use for storing the response from the Custom Resource.\n\n" +
+					"The IAM principal configured for the provider must have `s3:PutObject`, `s3:HeadObject` and `s3:GetObject` permissions on this bucket.",
+				TypeSpec: pschema.TypeSpec{Type: "string"},
+			},
+			"bucketKeyPrefix": {
+				Description: "The prefix to use for the bucket key when storing the response from the Custom Resource provider.",
+				TypeSpec:    pschema.TypeSpec{Type: "string"},
+			},
+			"serviceToken": {
+				Description: "The service token to use for the Custom Resource. The service token is invoked when the resource is created, updated, or deleted.\n" +
+					"This can be a Lambda Function ARN with optional version or alias identifiers.\n\n" +
+					"The IAM principal configured for the provider must have `lambda:InvokeFunction` permissions on this service token.",
+				TypeSpec: pschema.TypeSpec{Type: "string"},
+			},
+			"customResourceProperties": {
+				Description: "The properties to pass as an input to the Custom Resource.\nThe properties are passed as a map of key-value pairs whereas all " +
+					"primitive values (number, boolean) are converted to strings for CloudFormation interoperability.",
+				TypeSpec: pschema.TypeSpec{
+					Type: "object",
+					AdditionalProperties: &pschema.TypeSpec{
+						Ref: "pulumi.json#/Any",
+					},
+				},
+			},
+			"resourceType": {
+				Description: "The CloudFormation type of the Custom Resource. For example, `Custom::MyCustomResource`.\n" +
+					"This is required for CloudFormation interoperability.",
+				TypeSpec: pschema.TypeSpec{Type: "string"},
+			},
+			"stackId": {
+				Description: "A stand-in value for the CloudFormation stack ID. This is required for CloudFormation interoperability.\n" +
+					"If not provided, the Pulumi Stack ID is used.",
+				TypeSpec: pschema.TypeSpec{Type: "string"},
+			},
+		},
+		RequiredInputs: []string{"bucketName", "bucketKeyPrefix", "serviceToken", "customResourceProperties", "resourceType"},
+	}
+}
+
+type customResourceInvokeData struct {
+	event           cfn.Event
+	bucket          string
+	bucketKeyPrefix string
+	timeout         time.Duration
+	loggingLabel    string
+	serviceToken    string
+}
+
+// Check validates the inputs of the resource and applies default values if necessary.
+// It returns the inputs, validation failures, and an error if the inputs cannot be unmarshalled.
+func (c *cfnCustomResource) Check(ctx context.Context, urn urn.URN, randomSeed []byte, inputs resource.PropertyMap, state resource.PropertyMap, defaultTags map[string]string) (resource.PropertyMap, []ValidationFailure, error) {
```

Comment:
I *think* this might still receive unknowns but I'm not 100% sure.

Reply:
I'll double check. It's not a big deal if we stick to using the PropertyMap for Check